src/mondeath.cpp

```diff
@@ -625,6 +626,77 @@ void mdeath::kill_breathers(monster *z)
     }
 }
 
+void mdeath::detonate(monster *z)
+{
+    // How much ammo do we have left?
+    int ammo_count = 0;
+    for (auto am : z->ammo) {
+        ammo_count += am.second;
+    }
+    std::vector<std::string> pre_dets;
+    if (ammo_count > 0) {
+        // Pick three random ammo
+        for (int i = 0; i < 3; i++) {
```

Comment:
Oh, look: a weighted list in disguise. Just noting, so it can be converted to the shiny new generic implementation  of those, later.